onnx/defs/traditionalml/defs.cc

```diff
@@ -831,6 +836,26 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
             fail_shape_inference(
                 "Only one of the attributes 'base_values', 'base_values_as_tensor' should be specified.");
           }
+
+          // First axis of Y & Z - N (inputs)
```

Comment:
IMO, the code can be rewritten as:
```cc
checkInputRank(ctx, 0, 2);
Dim N, E;
unifyInputDim(ctx, 0, 0, N);
std::vector<int64_t> class_ids;
auto has_ids = getRepeatedAttribute(ctx, "class_ids", class_ids);
if (has_ids) {
   unifyDim(E, class_ids.size()); // or, just: E.set_dim_value(class_ids.size());
}
updateOutputShape(ctx, 0, {N});
updateOutputShape(ctx, 1, {N, E});
```

Reply:
Great suggestion! I wasn't aware of these higher-level functions. I applied it & refactored in the cases of the tree ensemble operators.